assets/aml-benchmark/components/src/prompt_crafter/package/prompt_factory.py

```diff
@@ -0,0 +1,340 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+"""PromptFactory class to create prompts from data."""
+
+
+import json
+import random
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Union
+from jinja2 import Environment
+
+from azureml._common._error_definition.azureml_error import AzureMLError
+
+from utils.exceptions import BenchmarkValidationException
+from utils.error_definitions import BenchmarkValidationError
+from .prompt import (
+    PromptType,
+    Prompt,
+    CompletionsPrompt,
+    ChatPrompt,
+    OpenAICreate,
+    OpenAICreateChatPrompt,
+    Role,
+)
+
+
+logger = logging.getLogger(__name__)
+
+PATIENCE = 100
+
+JINJA_ENV = Environment(keep_trailing_newline=True)
+
+
+@dataclass
+class PromptFactory(ABC):
+    """Factory for creating prompts."""
+
+    n_shots: int
+    prompt_pattern: str
+    few_shot_pool: Optional[List] = None
+    few_shot_pattern: Optional[str] = None
+    few_shot_separator: Optional[str] = None
+    prefix: Optional[str] = None
+    label_map_str: Optional[str] = None
+    output_pattern: Optional[str] = None
+    system_message: Optional[str] = None
+    metadata_keys: Optional[str] = None
+    additional_payload: Optional[str] = None
+    _messages: Optional[OpenAICreate] = None
+
+    def __post_init__(self):
+        """Validate and initialize the prompt factory."""
+        if self.n_shots > 0 and self.n_shots > len(self.few_shot_pool) and self.few_shot_pattern:
+            raise ValueError(f"n_shots ({self.n_shots}) > |few_shot pool| ({len(self.few_shot_pool)})")
+
+        self.label_map_jinja_prefix = ""
+        if self.label_map_str:
+            self.label_map_jinja_prefix = self._create_label_map_jinja_prefix(self.label_map_str)
+            self._inverse_label_map = self._inverse_label_map(self.label_map_str)
+
+        # Validate prompt_pattern
+        if self.prompt_pattern:
+            self.prompt_pattern = self.label_map_jinja_prefix + self.prompt_pattern
+            self.validate_jinja_template(self.prompt_pattern, "Prompt pattern is not a valid jinja pattern.")
+        else:
+            raise ValueError("A prompt pattern (in jinja template) is required.")
+
+        # Validate output_pattern
+        if self.output_pattern:
+            self.validate_jinja_template(self.output_pattern,
```

Comment:
Do we need to set `self.output_pattern = self.label_map_jinja_prefix + self.output_pattern`?

Reply:
I believe this was part of old code.